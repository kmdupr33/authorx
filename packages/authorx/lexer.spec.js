const { makeLexer } = require("./lexer");
const { stripIndentation } = require("./strip-indentation");
it("compiles without crashing", () => {
  makeLexer();
});
describe("lexer", () => {
  let lexer;
  let lex;
  beforeEach(() => {
    lexer = makeLexer();
    lex = (string) => {
      lexer.reset(string);
      const states = [];
      for (
        let token = lexer.next();
        token !== undefined;
        token = lexer.next()
      ) {
        states.push(token);
      }
      return states;
    };
  });
  it("lexes <p {hello}", () => {
    const tokens = lex("<p {hello}");
    expect(tokens).toMatchSnapshot();
  });
  it("lexes <p{hello <b{world}}", () => {
    const tokens = lex("<p{hello <b{world}}");
    expect(tokens).toMatchSnapshot();
  });
  it("lexes <p { <b { hello } world", () => {
    const tokens = lex("<p { <b { hello } world");
    expect(tokens).toMatchSnapshot();
  });
  it("lexes 1 \\< 2", () => {
    const tokens = lex("<p { 1 \\< 2 }");
    expect(tokens).toMatchSnapshot();
  });
  it("lexes multi argument invocations", () => {
    const tokens = lex("<c(r, exec) {}");
    expect(tokens).toMatchSnapshot();
  });
});
