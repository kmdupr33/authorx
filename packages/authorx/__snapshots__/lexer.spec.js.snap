// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer lexes <p { <b { hello } world 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "<",
    "toString": [Function],
    "type": "functionInvocation",
    "value": "<",
  },
  Object {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "p",
    "toString": [Function],
    "type": "identifier",
    "value": "p",
  },
  Object {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "{",
    "toString": [Function],
    "type": "brackets",
    "value": "{",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "<",
    "toString": [Function],
    "type": "functionInvocation",
    "value": "<",
  },
  Object {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "b",
    "toString": [Function],
    "type": "identifier",
    "value": "b",
  },
  Object {
    "col": 8,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": "{",
    "toString": [Function],
    "type": "brackets",
    "value": "{",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "hello ",
    "toString": [Function],
    "type": "words",
    "value": "hello ",
  },
  Object {
    "col": 17,
    "line": 1,
    "lineBreaks": 0,
    "offset": 16,
    "text": "}",
    "toString": [Function],
    "type": "brackets",
    "value": "}",
  },
  Object {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 19,
    "line": 1,
    "lineBreaks": 0,
    "offset": 18,
    "text": "world",
    "toString": [Function],
    "type": "words",
    "value": "world",
  },
]
`;

exports[`lexer lexes <p {hello} 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "<",
    "toString": [Function],
    "type": "functionInvocation",
    "value": "<",
  },
  Object {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "p",
    "toString": [Function],
    "type": "identifier",
    "value": "p",
  },
  Object {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": " ",
    "toString": [Function],
    "type": "ws",
    "value": " ",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "{",
    "toString": [Function],
    "type": "brackets",
    "value": "{",
  },
  Object {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "hello",
    "toString": [Function],
    "type": "words",
    "value": "hello",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "}",
    "toString": [Function],
    "type": "brackets",
    "value": "}",
  },
]
`;

exports[`lexer lexes <p{hello <b{world}} 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "<",
    "toString": [Function],
    "type": "functionInvocation",
    "value": "<",
  },
  Object {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "p",
    "toString": [Function],
    "type": "identifier",
    "value": "p",
  },
  Object {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "{",
    "toString": [Function],
    "type": "brackets",
    "value": "{",
  },
  Object {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "hello ",
    "toString": [Function],
    "type": "words",
    "value": "hello ",
  },
  Object {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "<",
    "toString": [Function],
    "type": "functionInvocation",
    "value": "<",
  },
  Object {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "b",
    "toString": [Function],
    "type": "identifier",
    "value": "b",
  },
  Object {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "{",
    "toString": [Function],
    "type": "brackets",
    "value": "{",
  },
  Object {
    "col": 13,
    "line": 1,
    "lineBreaks": 0,
    "offset": 12,
    "text": "world",
    "toString": [Function],
    "type": "words",
    "value": "world",
  },
  Object {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": "}",
    "toString": [Function],
    "type": "brackets",
    "value": "}",
  },
  Object {
    "col": 19,
    "line": 1,
    "lineBreaks": 0,
    "offset": 18,
    "text": "}",
    "toString": [Function],
    "type": "brackets",
    "value": "}",
  },
]
`;
